CE Comments resolvement

lukasaric opened the pull request:
This PR enables marking discussion as resolved.
- [x] Resolve comments functionality & preview resolved comments
- [x]  Unresolve comments from CE discussion

lukasaric committed:
Implement comments resolve :tada:

lukasaric committed:
Replaced actual comment with resolved label

lukasaric committed:
Extract discussion resolvement as component

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
UI tweaks :art:

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Unresolved comments in after cr8 hook

lukasaric committed:
Resolved btn & label UI tweaks :art:

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Resolve btn tweaks

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Minor component structure improvement

- Renamed resolvment component
- Moved it within discussion

lukasaric committed:
Minor condition correction :eyes:

lukasaric committed:
Cleanup :rocket:

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Used comments fetch after resolvement

lukasaric committed:
Comment Event listeners cleanup

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Added `editedAt` column

lukasaric committed:
Infer comment migrations :rocket:

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Added migration transaction

lukasaric committed:
Migration cleanup

lukasaric committed:
Merge branch 'feature/sidebar-ce-comments' into feature/ce-comments-resolvement

lukasaric committed:
Set paranoid true for comments listing

lukasaric committed:
CE discussion outside click adjustments

lukasaric committed:
Confirmation modal polish

lukasaric committed:
Confirmation modal enhancement :rocket:

lukasaric committed:
Resolve removed comments as well :bowtie:

underscope commented on server/shared/database/migrations/20210119091445-add-resolved-and-edited-at-column-to-comment.js:
```diff
@@ -0,0 +1,22 @@
+'use strict';
```
use `-` for splitting words and `_` for fields

underscope commented on server/shared/database/migrations/20210119091445-add-resolved-and-edited-at-column-to-comment.js:
```diff
@@ -0,0 +1,22 @@
+exports.up = async (qi, Sequelize) => {
+  const columns = getColumns(Sequelize);
+  const transaction = await qi.sequelize.transaction();
+  await Promise.all(columns.map(({ name, ...options }) => {
```
Tend to do these one by one ;)

lukasaric replied:
Hmm, I think It will be better to decouple this in 2 separate migrations to avoid using `bluebird` injection (to handle promises in series), and also to avoid using transaction. Also it's more readable (code, file naming..)

underscope commented on server/shared/database/migrations/20210119091445-add-resolved-and-edited-at-column-to-comment.js:
```diff
@@ -0,0 +1,22 @@
+
+exports.up = async (qi, Sequelize) => {
+  const columns = getColumns(Sequelize);
+  const transaction = await qi.sequelize.transaction();
```
Not sure if this is working since not all DBs support transactions with DDL. Please validate within docs.

lukasaric replied:
See above comment ☝️

underscope commented on client/components/common/tce-core/Events/DiscussionEvent.js:
```diff
@@ -1,5 +1,6 @@
   REMOVE: 'comment:remove',
-  SET_LAST_SEEN: 'comment:setLastSeen'
+  SET_LAST_SEEN: 'comment:setLastSeen',
+  RESOLVE: 'comments:resolve'
```
Isn't this element event ? :resolveComments. Can we do this for activities?

lukasaric replied:
It will be handled in a separate PR 😉

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -76,7 +76,8 @@ export default {
         const lastSeen = max([seen.contentElement[it.uid], seen.activity[activityUid]]);
-        acc[it.uid] = { ...it, comments, lastSeen: lastSeen || 0 };
+        const isResolved = !comments.length;
+        acc[it.uid] = { ...it, comments, isResolved, lastSeen: lastSeen || 0 };
```
It doesn't feel like isResolved should be attached like this directly on content element without notion of comments.  If you gonna go down that path consider something like hasUnresolvedComments

underscope commented on client/components/common/ConfirmationModal.vue:
```diff
@@ -36,13 +36,19 @@ export default {
+    show(val) {
+      const action = val ? this.context.onOpen : this.context.onClose;
+      if (!val) this.context = createContext();
+      return action && action();
```
why not keeping the initial version and adding onOpen onClose in open/close methods ? Seems more readable (no if for context reset, no action resolution) and does not create extra watcher 🤔

lukasaric replied:
I was going to do this initially then I realised clicking outside does not invoke close method. 
So it needs to stay like this unless you have any other idea?

lukasaric replied:
Nvm. Handled this with vuetify dialog `click:outside` event. 🤡

underscope commented on client/components/common/mixins/commentEventListeners.js:
```diff
@@ -2,34 +2,39 @@ import { mapActions, mapMutations } from 'vuex';
+      const { id, contentElementId: elementId } = comment;
+      const action = id ? 'updateComment' : 'saveComment';
+      await this[action]({ ...comment, activityId: this.activityId });
+      if (!hasComments) this.fetchComments({ elementId });
```
Not sure if this logic is self explanatory

underscope commented on client/components/common/tce-core/Discussion/ResolveButton.vue:
```diff
@@ -0,0 +1,38 @@
+<template>
+  <div class="resolve-button">
```
resolve-button-container ? Is this class used?

lukasaric replied:
Yes it is. It's controlled outside within parent

underscope commented on client/components/common/tce-core/Discussion/ResolveButton.vue:
```diff
@@ -0,0 +1,38 @@
+          Resolve Discussion
+        </v-btn>
+      </template>
+      <span>Mark discussion as resolved</span>
```
Mark as resolved and hide discussion

underscope commented on client/components/common/tce-core/Discussion/ResolveButton.vue:
```diff
@@ -0,0 +1,38 @@
+import events from '../Events/DiscussionEvent';
+
+export default {
+  name: 'resolve-button',
```
A bit too generic name: resolve-comments(-btn) ?

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -87,32 +93,33 @@ export default {
       this.editorBus.emit(events.SET_LAST_SEEN, options);
     }
   },
+  provide() {
```
Why not just emit event from child ? 🤔

lukasaric committed:
Decoupled comment columns migrations

- ..to avoid using transactions & bluebird injection

lukasaric committed:
Naming & text cleanup :eyes:

lukasaric committed:
Set resolve as functional comp. & events polish

lukasaric committed:
Unresolved comments flag naming improvement

lukasaric committed:
Confirmation event actions enhancement :surfer:

lukasaric committed:
Prevent calling unexisting onOpen/close function

lukasaric committed:
Confirmation cleanup :lipstick:

kronicker commented on client/components/common/mixins/commentEventListeners.js:
```diff
@@ -2,34 +2,39 @@ import { mapActions, mapMutations } from 'vuex';
     }
   },
   mounted() {
     COMMENT_EVENTS.forEach(({ event, action }) => {
```
Is there a reason to do this on `mounted` instead `created`?

kronicker commented on client/components/common/tce-core/Discussion/ResolveButton.vue:
```diff
@@ -0,0 +1,26 @@
+      <template v-slot:activator="{ on }">
+        <v-btn
+          v-on="on"
+          @click="listeners.resolve"
```
Why not just proxy listeners here? Since this is just a button I think keeping `click` makes more sense than having `resolve` event.

kronicker commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -1,5 +1,6 @@
 <template>
   <div ref="discussion" class="embedded-discussion">
+    <resolve-button v-if="showResolveButton" v-on="$listeners" />
```
Aligning with the previous comment I would remove this listeners proxy in favor of `@click="$emit('resolve')"`

kronicker commented on client/components/common/tce-core/Discussion/Thread/Comment/Header.vue:
```diff
@@ -62,9 +62,10 @@ export default {
     isDeleted: vm => !!vm.comment.deletedAt,
-    showOptions: vm => vm.isAuthor && !vm.isDeleted,
+    isResolved: vm => vm.comment.resolved,
+    showEditedLabel: vm => !!vm.comment.editedAt && !vm.isResolved,
```
Why not show the "edited" label for resolved comments?

kronicker commented on server/shared/database/migrations/20210125094819-add-resolved-column-to-comment.js:
```diff
@@ -0,0 +1,11 @@
+const COLUMN_NAME = 'resolved';
+
+exports.up = (qi, { BOOLEAN }) => qi.addColumn(TABLE_NAME, COLUMN_NAME, {
+  type: BOOLEAN,
```
What do you think about making this `resolvedAt` timestamp instead of Boolean. Same reasoning as for `editedAt`.

kronicker commented on server/comment/comment.controller.js:
```diff
@@ -40,9 +41,16 @@ function remove({ comment }, res) {
 }
 
+async function resolve({ body }, res) {
+  const where = { contentElementId: body.contentElementId };
```
You should guard this API against `contentElementId` being `null`

kronicker commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -3,7 +3,8 @@ import get from 'lodash/get';
 
 export const getComments = state => params => {
-  const comments = filter(state.items, params);
+  const opts = params.contentElementId ? { ...params, resolved: false } : params;
```
Let's make this more readable like:
```js
if (params.contentElementId) params.resolved = false;
```

kronicker commented on client/store/modules/repository/comments/actions.js:
```diff
@@ -11,11 +11,16 @@ const plugSSE = ({ commit }) => {
 };
 
-const fetch = ({ commit }, payload) => {
+const fetch = ({ commit }, payload = {}) => {
```
No need for defaulting here?

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Cast missing `isResolved` to boolean

lukasaric committed:
Throw err if element id does not exist

kronicker reviewed (approved):
@underscope LGTM 👍 

@kjuej @markolovric We could use some QA eyes on this.

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -76,7 +76,8 @@ export default {
         const comments = this.getComments({ activityId, contentElementId: it.id });
         const lastSeen = max([seen.contentElement[it.uid], seen.activity[activityUid]]);
-        acc[it.uid] = { ...it, comments, lastSeen: lastSeen || 0 };
+        const hasUnresolvedComments = !comments.length;
```
missing one `!` (since getComments is returning unresolved comments) ? 🤔 
@lukasaric @kronicker ?

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -3,6 +3,7 @@ import get from 'lodash/get';
 import orderBy from 'lodash/orderBy';
 
 export const getComments = state => params => {
+  if (params.contentElementId) params.resolvedAt = null;
```
Maybe name should reflect the nature or returning unresolved comments or resolve controlling arg should be passed 🤔

bubafinder commented:
We need to cover comments for deleted elements. Maybe to introduce secondary color over element type and add tooltip "Element deleted"

![image](https://user-images.githubusercontent.com/41724080/106582558-67e25600-6544-11eb-96c9-e80fade18723.png)

bubafinder commented:
Update on the last comment, after refreshing the page, the element type isn't visible anymore.
Maybe to make this move reactive?
We could definitely consider leaving a red element type chip with an explanation tooltip.

bubafinder commented:
Expand container width for bigger comments

![image](https://user-images.githubusercontent.com/41724080/106584290-4edaa480-6546-11eb-9c0f-f1d655d28a0c.png)

bubafinder commented:
Add scroll feature inside comments element.

![image](https://user-images.githubusercontent.com/41724080/106585751-f6a4a200-6547-11eb-84b7-0c602c16bb99.png)

bubafinder commented:
Make resolvment also reactive as new comment. In this sample, another user1 already resolved this conversation, but user2 cant see it yet after opening conversation.

![image](https://user-images.githubusercontent.com/41724080/106599434-56577900-6559-11eb-8e12-5f8b5ff96a85.png)

bubafinder commented:
When comment window is opened, move parent content element in selected state. In this sample, focus is still on the upper content element.

![image](https://user-images.githubusercontent.com/41724080/106599823-e4cbfa80-6559-11eb-8eac-618b4dfa44d9.png)

bubafinder commented:
Prioritize First&Last name over email. In this case, this user has setted First and Last name in Profile page.

![image](https://user-images.githubusercontent.com/41724080/106600905-4e003d80-655b-11eb-9413-f4008ca80022.png)

lukasaric committed:
Cleanup :lipstick:

- Added scrolling overflow to comments thread
- Fix user attrs
- hasUnresolved flag fixed

lukasaric committed:
Set in motion SSE for resolvement

lukasaric committed:
Comment SSE adjustment after deleted element

lukasaric committed:
Added scrolling only for CE comments

lukasaric committed:
Hide vertical scrollbar on CE thread

lukasaric committed:
Added single comment resolvement

lukasaric committed:
Resolvement enhancement :rocket:

- Added single unresolvement
- Added preview component
- Removed common text editor
- UI tweaks :art:

lukasaric committed:
Detached comment from activity after CE deletion

lukasaric committed:
Comment header options cleanup

lukasaric committed:
typo :eyes:

lukasaric committed:
Reused resolve flag from comment index

lukasaric committed:
Class naming :eyes:

lukasaric committed:
Added after bulk update hook

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Added repository id in comment resolve where

lukasaric committed:
Added confirmation for resolve all & UI tweaks

lukasaric committed:
Added unresolve comment tooltip
